### ISSUE TITLE
Math support in NeXus

### ISSUE THREAD

**Issue by phyy-nx** [nexusformat/definitions#711]
Hi folks, I know this opens up a topic previously closed (see #376), but I think it's worth revisiting.  We have two detectors, the Jungfrau 16M at SwissFEL and the AGIPD detector at EuXFEL, that both have complex calibration procedures.  These detectors have very high data rates and it is costly both in run-time at the beamline and in storage space to apply the calibrations in memory and then write out the data to disk (for example, the calibration converts from int16 to float64).  It would be better to instead have the master file link to the raw data and the gain and pedestal maps, and then specify how the corrections are applied by an equation.

A simplified version of the equation is
```
calibrated_data = (raw_data-pedestal)/gain
```
(this is complicated by the bitwise operators needed to select between the 3 gains)

So, as part of the code camp 2020, we're opening this issue back up.  How could we build NeXus files that include raw data and the instructions on how to apply the calibrations?  What is the language that the equation could be written in?  Would we need to look into parsers ourselves, if so what are available?

(passes ball to @benajamin)

**Comment by FilipeMaia** [comment]
I think this is a very interesting subject, but one should be careful not to make things too complex.
I wonder if the same result could be achieved with a smart HDF5 filter, instead of handling it explicitly in NeXus?

**Comment by benajamin** [comment]
Yes, this issue has a strong use-case, but is fraught with problems. Certainly, executable code in data files could be a security nightmare, especially if it isn't heavily restricted. My previous research found a few maths parsers that could be useful since they both limit what the "embedded code" can do and promise good performance in crunching the numbers. Here are some maths parsers I found:

Name  | Implemented | License | Bindings/translations
-- | -- | -- | -- 
[MuParser](https://beltoforion.de/article.php?a=muparser)  | C++ | MIT | [python](https://github.com/Ayuto/Py-MuParser)
[MuParserX](https://beltoforion.de/article.php?a=muparserx)  | C++ | BSD | 
[mXparser](http://mathparser.org/) | JAVA, C#.NET | Simplified BSD  |   
[ExprTk](http://www.partow.net/programming/exprtk/index.html) | C++ | MIT | [python](https://pypi.org/project/cexprtk/)
[TinyExpr](https://github.com/codeplea/tinyexpr) | ANSI C | zlib |   |  
[JS Expr-Eval](https://github.com/silentmatt/expr-eval) | Javascript | MIT | [python](https://github.com/AxiaCore/py-expression-eval)

The next step would be to try actually making use of one (or more) of these parsers to make sure it is a viable solution to your problem.

**Comment by benajamin** [comment]
For a NeXus solution, we would need the following ingredients:
* a string containing the math equation following a dictionary of allowed syntax, operators and functions
* a set of links between equation variables and HDF5 datasets/attributes to take values from
* a location to store the result
* a flag to indicate (or method to determine) if the calculation has been performed yet.

We could use a [null-dataset](http://docs.h5py.org/en/stable/high/dataset.html#creating-and-reading-empty-or-null-datasets-and-attributes) for the location to store the result and provide the equation string and set of variable-pointers as attributes to this null-dataset. A successful evaluation would then replace the null-dataset with the resulting array of values. For example a file could contain:

/entry/instrument/detector/raw_data = _array_
/entry/instrument/detector/offset = _scalar or array_
/entry/instrument/detector/gain = _scalar or array_
/entry/instrument/detector/data = _null_
/entry/instrument/detector/data@evaluate = "(A-B)/C"
/entry/instrument/detector/data@var_A = "/entry/instrument/detector/raw_data"
/entry/instrument/detector/data@var_B = "/entry/instrument/detector/offset"
/entry/instrument/detector/data@var_C = "/entry/instrument/detector/gain"

**Comment by PeterC-DLS** [comment]
There was mention of using TeX. A quick search brings up:
* https://github.com/augustt198/latex2sympy
* https://tex.stackexchange.com/questions/83731/how-to-evaluate-an-expression-before-printing-it
* https://tex.stackexchange.com/questions/4223/what-parsers-for-latex-mathematics-exist-outside-of-the-tex-engines

**Comment by PeterC-DLS** [comment]
There's a new game in town: https://github.com/lucasvr/hdf5-udf

**Comment by phyy-nx** [comment]
From NIAC 2020:

- UDF looks interesting.  Need to investigate security model.  Need to ask the HDF5 group about long-term UDF support.  What version of HDF5 supports UDF?
- MuParser: how does this interact with numpy/h5py to be performant?
- Is it possible to combine a MuParser application into a plugin that the NIAC could support?
- What additions do we need to add to NXdetector to specify gain and pedestal datasets?  Or would we add no additions as the equations themselves would specify the location of the gain and pedestal datasets?

@phyy-nx will investigate.

**Comment by phyy-nx** [comment]
From Telco Dec 20 2023, we would like to form a subcommittee to resolve issues related to math support in NeXus.  Current members @domna @benajamin @PeterC-DLS @yayahjb.  Please distribute your discussions!  We'd like to resolve this by NIAC 2024 (Sept 2024).

**Comment by FilipeMaia** [comment]
FYI here's an example of what could be done with an HDF5 filter, https://github.com/FilipeMaia/h5calib

**Comment by phyy-nx** [comment]
@FilipeMaia a filter like this could be (or maybe is) a conda package, right?

**Comment by FilipeMaia** [comment]
That repository was just a test to see if it could be done. But it could certainly be more carefully developed and turned into a conda package.

**Comment by phyy-nx** [comment]
👍

**Comment by woutdenolf** [comment]
My opinion on the math/formula discussion: do not introduce string based types/fields meant for anything else than documentation. Meaning it is not intended to be evaluated, parsed, validated or dereferenced by a machine. Also validators and parsers can be exploited by a carefully crafted NX_CHAR.

**Comment by phyy-nx** [comment]
Proposal, three options:

Option 1)
Users may express formulas in NX_CHAR.  It is assumed they will not be evaluated.

Option 2)  NeXus will support math using the following steps:

1. Choose a grammar.  Requested features include standard arithmetic, log, exponentiation, and bit shift
2. Add new type, NX_MATH, a formula string
3. Define how symbols are used, which are fields in the same group as the NX_MATH field.  Use links as needed.
4. In the NX_MATH documentation provide lots of warnings, including that software should not directly evaluate NX_MATH fields without checking the expression matches the grammar exactly
5. Provide a list of software packages that implement at least checking the grammar and maybe executing it.  Include the major programming languages (Python, C++, R, Javascript)

Option 3)
NeXus shall not be code.  It is permissible to describe calibration procedures, but not to include formulas.

Option 4) Punt decision for more research and debate.  FairMAT will use formula descriptions instead of formulas.

Option 4 is adopted by vote in [NIAC 2024](https://www.nexusformat.org/NIAC2024.html)

(leaving issue open for now)

**Comment by benajamin** [comment]
My research on this topic indicates that including any kind of executable code (or expectations of it) in a data file is very dangerous. Since we do not control the code that will read the files, and we expect the files to be processed in places that are considered high value targets for hacking, any useful version of math support in NeXus is going to cause big trouble sooner or later. 

I would rather not implement math support in order to avoid needing to scan nexus files for malicious code.

**Comment by FilipeMaia** [comment]
Having arbitrary executable code is indeed very dangerous. But limiting it to a pure data-flow operator, without control-flow jumps, shared mutable state, or side effects, could make it manageable.

**Comment by phyy-nx** [comment]
Without math support, I cannot describe how raw data should be calibrated by downstream programs, which means all programs must be custom instead of general.  I call back to my opening argument about the Jungfrau 16M detector, which has three gain modes.  It's not really this:

```
calibrated_data = (raw_data-pedestal)/gain
```
Which you could argue is actually currently supported, due to [data_scaling_factor](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-data-data-scaling-factor-field) and [data_offset](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-data-data-offset-field), which can sub in for gain and pedestal, respectively.  However, it's really more like this:

```
calibrated_data = (((raw_data-pedestal1)/gain1) & mask1) |
                  (((raw_data-pedestal2)/gain2) & mask2) |
                  (((raw_data-pedestal3)/gain3) & mask3)
```

Where the `&` and `|` operators are bitwise.  Right now, because we can't represent the calibration procedure in the metadata, either the raw data (16 bit int) is calibrated in memory using custom, hardcoded functions specific to this detector, or the data is calibrated when it is produced and saved as float32 or float64, greatly increasing file size (by far the more common approach).

I think Option 2 [above](https://github.com/nexusformat/definitions/issues/711#issuecomment-2383124422) is a careful, responsible approach to allowing representation of mathematics. However, I appreciate it's been controversial, hence the age of this issue :)
